common/src/primitives/signed_amount.rs

```diff
@@ -0,0 +1,1030 @@
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(clippy::eq_op)]
+
+use std::iter::Sum;
+
+use super::Amount;
+
+// use only unsigned types
+// if you need a signed amount, we should create a separate type for it and implement proper conversion
+pub type SignedIntType = i128;
+
+/// An unsigned fixed-point type for amounts
+/// The smallest unit of count is called an atom
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SignedAmount {
+    val: SignedIntType,
+}
+
+fn remove_right_most_zeros_and_decimal_point(s: String) -> String {
+    let point_pos = s.chars().position(|c| c == '.');
+    if point_pos.is_none() {
+        return s;
+    }
+    let s = s.trim_end_matches('0');
+    let s = s.trim_end_matches('.');
+    s.to_owned()
+}
+
+impl SignedAmount {
+    pub const MAX: Self = Self::from_atoms(SignedIntType::MAX);
+
+    pub const fn from_atoms(v: SignedIntType) -> Self {
+        SignedAmount { val: v }
+    }
+
+    pub fn into_atoms(&self) -> SignedIntType {
+        self.val
+    }
+
+    pub fn from_unsigned(amount: Amount) -> Option<Self> {
+        let unsigned_atoms = amount.into_atoms();
+        let atoms: SignedIntType = unsigned_atoms.try_into().ok()?;
+        Some(Self::from_atoms(atoms))
+    }
+
+    pub fn into_unsigned(self) -> Option<Amount> {
+        let atoms = self.val;
+        let unsigned_atoms: super::amount::UnsignedIntType = atoms.try_into().ok()?;
+        Some(Amount::from_atoms(unsigned_atoms))
+    }
+
+    pub fn into_fixedpoint_str(self, decimals: u8) -> String {
+        let amount_str = self.val.abs().to_string();
+        let decimals = decimals as usize;
+        let sign = if self.val < 0 { "-" } else { "" };
+        if amount_str.len() <= decimals {
+            let zeros = "0".repeat(decimals - amount_str.len());
+            let result = sign.to_owned() + "0." + &zeros + &amount_str;
+
+            remove_right_most_zeros_and_decimal_point(result)
+        } else {
+            let ten: SignedIntType = 10;
+            let unit = ten.pow(decimals as u32);
+            let whole = self.val.abs() / unit;
+            let fraction = self.val.abs() % unit;
+            let result = format!("{sign}{whole}.{fraction:00$}", decimals as usize);
+
+            remove_right_most_zeros_and_decimal_point(result)
+        }
+    }
+
+    pub fn from_fixedpoint_str(amount_str: &str, decimals: u8) -> Option<Self> {
```

Comment:
Are conversions from strings like "0.8" or even better ".8" legit? I believe these cases are not uncovered by tests

Reply:
Done